opacus/optimizers/optimizer.py

```diff
@@ -167,13 +163,12 @@ def pre_step(self) -> Optional[float]:
             self.step_hook(self)
 
         self._is_last_step_skipped = False
+        return True
 
     def step(self, closure: Optional[Callable[[], float]] = None) -> Optional[float]:
-        self.pre_step()
-        
-        return self.optimizer.step(closure)
-
-
-    def pre_step(self):
-        self.clip_and_accumulate()
+        if self.pre_step():
+            return self.optimizer.step(closure)
+        else:
+            return None
```

Comment:
There is probably a bug here. 

We do `self.clip_and_accumulate()` before the model is reevaluated by calling the closure, we should not have appropriate sample_grads to work with. This breaks optimization loop when using Lightning's [automatic optimization](https://pytorch-lightning.readthedocs.io/en/latest/common/optimizers.html#automatic-optimization). 

This may work incorrectly with the manual optimization loop in our examples: we must be using grad_samples from the previous iteration, not the current one. I'm not sure though.

I tried to fix this for the old PrivacyEngine [here](https://github.com/pytorch/opacus/pull/244/files#r744205844).

Reply:
ouch
tbh, it is true that `closure` was an afterthought here. I wasn't familiar with the applications you've listed in your PR.

would simply calling `closure()` before `clip_and_accumulate()` do the trick?

Reply:
Adding this in the beginning of `step()` should partially solve the problem:

```python
                loss = None
                if closure is not None:
                    with torch.enable_grad():
                        loss = closure()
```

In the end we need to return `loss` instead of `self.optimizer.step(closure)` output. 

However, [I don't know what happens](https://github.com/pytorch/opacus/pull/244/files#r744265353) in case the optimizer makes several backward passes during one step. It is not clear to me if it can break DP accounting. 

I need to investigate this. For this PR I would either merge it as is or add my fix, but in any case leave a comment / TODO.